### ISSUE TITLE
[HOLD for payment 2023-04-20] [$4000] Using a header (#) and a space or bold (*), then adding a code block results in the code block being bolded for a moment

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15193]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open a chat
2. Start a message with # then a space
3. Add a code block using three backticks

You should see the code block go bold, then normal.

## Expected Result:
A code block should not show bold text.

## Actual Result:
Bold text appears temporarily, then reverts to normal text.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.72-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/219246458-0b133e30-c65b-49bf-a7c4-d631f67d62c1.mov

https://user-images.githubusercontent.com/43996225/219246502-d75faa2f-fe7c-4105-bc62-bedbc19de6d6.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @gadhiyamanan 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1672998135151449

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014c5422ceeae647ae</li>
        <li>Upwork Job ID: 1626059997054926848</li>
        <li>Last Price Increase: 2023-03-13</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jliexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by jliexpensify** [comment]
Can reproduce in v1.2.72-0 as well. Seems to be occurring for Android (v1.2.71-1) as well.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~014c5422ceeae647ae

**Comment by MelvinBot** [comment]
Current assignee @jliexpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @aimane-chnaif (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @roryabraham (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Send 

```text
# ```test```
```

will be display in bold and later be normal.

### What is the root cause of that problem?

Frontend translates

```
# ```test```
```

to html

```
<h1><pre>test</pre></h1>
```

So the message will dispayed in bold first and later backend return result html

```
<h1></h1><pre>test</pre>
```

So the message displays normally later

### What changes do you think we should make in order to solve the problem?

To fix this issue we need to improve ExpensiMark to translate markdown header to html `h1` tag here

https://github.com/Expensify/expensify-common/blob/4f61e5529c89936bfb72b6cee53250990c16f188/lib/ExpensiMark.js#L166

to consider not to match some html tags, like `(?![^<]*(<\/pre>|<\/code>|<\/a>))`

We also need to sync backend solution to frontend to ensure consistency.

### What alternative solutions did you explore? (Optional)

None

**Comment by eh2077** [comment]
Just tested using GitHub comment. It should be displayed in bold right?

**Comment by jliexpensify** [comment]
@eh2077 - code blocks [shouldn't be bold at all](https://expensify.slack.com/archives/C049HHMV9SM/p1676480283232819?thread_ts=1672998135.151449&cid=C049HHMV9SM).

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Send 

```text
# ```test```
```

will be display in bold and later be normal.

### What is the root cause of that problem?

Frontend translates

```
# ```test```
```

to html

```
<h1><pre>test</pre></h1>
```

So the message will dispayed in bold first and later backend return result html

```
<h1></h1><pre>test</pre>
```

So the message displays normally later

### What changes do you think we should make in order to solve the problem?

To fix this issue we need to improve ExpensiMark to translate markdown header to html `h1` tag here

https://github.com/Expensify/expensify-common/blob/4f61e5529c89936bfb72b6cee53250990c16f188/lib/ExpensiMark.js#L166

to consider not to match some html tags, like 

```typescript
/^#(?!#) +([\s\S]*?)(?![^<].*(<\/pre>|<\/code>))/gm,
```

We also need to sync backend solution to frontend to ensure consistency.

### What alternative solutions did you explore? (Optional)

None

**Comment by dgaisan** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Bold text appears temporarily, then reverts to normal text when sending `# ```to send``` ` message

### What is the root cause of the problem?
When the message is synched with the backend it's being changed from the original one:
- original message: ` # bold code `
- The message gets HTML-escaped before synching with the backend: ` <h1><pre>bold code<pre></h1> `
- Then message is send to the backend, and after being processed on the backend it looks as following: ` <h1></h1><pre>bold code</pre> `

Therefore the text flickering (from bold to regular) happens due to this change on the backend.
This can be clearly seen in offline mode: the message stays bold forever, until the app goes back online  

### What changes do you think we should make in order to solve the problem?
To make sure everything is consistent I suggest to simply get rid of Bold property in the message before it's even rendered for the first time. 
The change can be made [in this method](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionCompose.js#L471)
The advantage of this solution is the fact that it's going to be very clean, we won't need to make changes on the backend, and we'll maintain consistency w/o introducing any regressions...

### What alternative solutions did you explore? (Optional)
None

**Comment by allroundexperts** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Message turns bold temporarily when enclosed with a `#` followed by ```.

### What is the root cause of that problem?
The frontend parses the expression in a different way then the backend. Frontend encloses the `<pre>` tag inside the `h1` tag while the backend does not enclose the `pre` tag inside `h1`.

### What changes do you think we should make in order to solve the problem?

We can change the ExpensiMark `#` parsing rule such that it ignores adding `h1` if `pre` or `code` tag is present.
The rule exists here:
https://github.com/Expensify/expensify-common/blob/4f61e5529c89936bfb72b6cee53250990c16f188/lib/ExpensiMark.js#L166

The new regex would look like the following:
```
^#(?!#) +(?!<(pre|code)>)(.+)$\s*
```

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

Ignoring the `#` if any tag is present after it. I think we don't want to do that. We should ignore only if `code` or `pre` tag is present.

**Comment by allroundexperts** [comment]
@Santhosh-Sellavel Here's how it works on Slack. IMO, it makes more sense to not allow this behaviour.

<img width="798" alt="Screenshot 2023-02-17 at 10 41 23 PM" src="https://user-images.githubusercontent.com/30054992/219730286-49461acc-09c1-41ef-87ac-2162860a718f.png">

**Comment by Santhosh-Sellavel** [comment]
@allroundexperts Slack doesn't support `#` styling text as a header as we do. So the behavior aligns with what GitHub does i.e support # for header and code or code blocks

**Comment by allroundexperts** [comment]
> @allroundexperts Slack doesn't support `#` styling text as a header as we do. So the behavior aligns with what GitHub does i.e support # for header and code or code blocks

You're right. My bad.

**Comment by jliexpensify** [comment]
Not overdue, still waiting for proposals.

**Comment by moezh** [comment]
[fix-15193](https://github.com/Expensify/App/compare/main...moezh:expensify-app:issue-15193)

In src/styles/styles.js (line 103): 
```diff
pre: {
...baseCodeTagStyles,
paddingTop: 12,
paddingBottom: 12,
paddingRight: 8,
paddingLeft: 8,
fontFamily: fontFamily.MONOSPACE,
marginTop: 0,
marginBottom: 0,
+ fontWeight: 'normal',
+ fontSize: variables.fontSizeNormal,
+ lineHeight: variables.fontSizeNormalHeight,
},
```

In src/styles/styles.js (line 133): 
```diff
h1: {
fontSize: variables.fontSizeLarge,
- marginBottom: 8,
+ marginBottom: 0,
+ marginTop: 0,
+ lineHeight: variables.lineHeightLarge,
},
```

**Comment by MelvinBot** [comment]
📣 @moezh! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Send 

```text
# ```test```
```

will be displayed in bold and later be normal.

### What is the root cause of that problem?

Frontend translates

```
# ```test```
```

to html

```
<h1><pre>test</pre></h1>
```

So the message will dispayed in bold first and later backend return result html

```
<h1></h1><pre>test</pre>
```

So the message displays normally later.

### What changes do you think we should make in order to solve the problem?

To fix this issue, we need to check if the match group of markdown heading contains HTML code fence. We can improve the replacement function

https://github.com/Expensify/expensify-common/blob/a2942d035614bf4bed90ffe2114d6ee828ee96da/lib/ExpensiMark.js#L168

to check if the heading contains HTML code fence tag, like 

```typescript
replacement: (__, g1) => {
    const index = g1.indexOf('<pre>');
    if (index === -1) {
        return `<h1>${g1}</h1>`;
    } else if (index === 0) {
        // heading is empty and keep text #
        return `# ${g1}`;
    } else {
        // heading is not empty and cut after the tag
        return `<h1>${g1.substr(0, index)}</h1>${g1.substr(index)}`;
    }
}
```

With the improvement above, markdown 

```
# ```test```
```

will be translated to an empty `<h1>` heading and a code fence, which is same as the result returned from backend

```
<h1></h1><pre>test</pre>
```

We can add following test cases to verify the solution

```
# ```test```
```

```
# Removing ```code``` blocks
```

### What alternative solutions did you explore? (Optional)

None

**Comment by moezh** [comment]
Contributor details
Your Expensify account email: moez.hachicha@icloud.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0189419ffb4bcfa145

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by Santhosh-Sellavel** [comment]
> @Santhosh-Sellavel Whether the expected behavior is bolded or not, this should be at least inconsistency bug between frontend and backend.


Yes that inconsistency is the bug IMO text should be bold I guess as it align with github. But let's wait for @roryabraham inputs too!

**Comment by thienlnam** [comment]
In that case, the code block shouldn't be bolded

**Comment by roryabraham** [comment]
> should we fix both header (#) and bold (*) cases in this GH?

Ideally, yes

**Comment by eh2077** [comment]
### Proposal 
[Updated](https://github.com/Expensify/App/issues/15193#issuecomment-1441378302)

Below is the solution to fix bold case.
### What changes do you think we should make in order to solve the problem?
To fix the issue for bold markdown, we can add a function to modify text including code fence tag, like
```typescript
modifyCodeFenceForTag(tagName, textToCheck) {
    const openTag = `<${tagName}>`;
    const closeTag = `</${tagName}>`;

    const textSplitted = textToCheck.split('<pre>');
    let joinedText = '';
    for (let i = 0; i < textSplitted.length; i++) {
        let text = '';
        if (Str.contains(textSplitted[i], '</pre>')) {
            const splitText = textSplitted[i].split('</pre>');
            if (splitText[1].trim() === '') {
                // retain spaces, like 'start<pre>code</pre>   <pre>code2</pre>'
                text = `<pre>${splitText[0]}</pre>${splitText[1]}`;
            } else {
                // retain leading and trailing spaces, like 'start<pre>code></pre>    middle    <pre>code2</pre>'
                const trimmedText = splitText[1].trim();
                text = `<pre>${splitText[0]}</pre>${Str.cutAfter(splitText[1], trimmedText)}${openTag}${trimmedText}${closeTag}${Str.cutBefore(splitText[1], trimmedText)}`;
            }
        } else if (textSplitted[i].trim() !== '') {
            // retain trailing spaces, like 'start     <pre>code</pre>'
            text = `${openTag}${textSplitted[i].trim()}${closeTag}${Str.cutBefore(textSplitted[i], textSplitted[i].trim())}`;
        }
        joinedText += text;
    }
    return joinedText;
}
```
And change this [line](https://github.com/Expensify/expensify-common/blob/fff304449214480065bf550e562ad39e681f6f95/lib/ExpensiMark.js#L145) to 
```js
replacement: (match, g1) => this.modifyCodeFenceForTag('strong', g1),
```
We can do the same fix for italic and strikethrough markdown as well if we can confirm them from product team.

**Comment by eh2077** [comment]
Hi @aimane-chnaif , thanks for looking at my proposal!

I just updated my proposal with plain English and pseudo code and would like to seek your review again, thanks!

**Comment by MelvinBot** [comment]
@jliexpensify @roryabraham @aimane-chnaif this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by jliexpensify** [comment]
^ Melvin, Aimane is reviewing proposals!

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $2000

**Comment by jliexpensify** [comment]
Hi everyone + @aimane-chnaif - I've been chatting to @roryabraham about this issue, and the fact that Melvin has been auto-doubling unnecessarily.

We've made the decision to drop this back down to $2000 for now, as there's currently a lot of healthy discussion and some solid proposals. Thanks for your understanding!

**Comment by jliexpensify** [comment]
Ah, good catch - sorry: that was a typo when I copied and pasted haha.

**Comment by eh2077** [comment]
> @eh2077 please test these cases and share demo video:
> 
> <img alt="test" width="270" src="https://user-images.githubusercontent.com/96077027/222633056-40432b6e-4724-4c2a-a1cf-08f496327123.png">

@aimane-chnaif 
Sure! It's a nice test case 👍 Below is the demo video

https://user-images.githubusercontent.com/117511920/222675669-60903949-ae45-43dd-845e-a20c81bb37ec.mov

Please note that I updated the [header proposal](https://github.com/Expensify/App/issues/15193#issuecomment-1441378302) a little bit to handle suffix spaces when the markdown heading includes code fence tag. As it's a slight change of the proposed solution, so I just fixed it.

**Comment by MelvinBot** [comment]
@jliexpensify, @roryabraham, @aimane-chnaif Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by eh2077** [comment]
Hi @aimane-chnaif , here's the diff for your information

```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index c3decc6..e54e157 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -140,7 +140,7 @@ export default class ExpensiMark {
                 // for * and ~: https://www.rexegg.com/regex-boundaries.html#notb
                 name: 'bold',
                 regex: /\B\*((?=\S)(([^\s*]|\s(?!\*))+?))\*\B(?![^<]*(<\/pre>|<\/code>|<\/a>))/g,
-                replacement: '<strong>$1</strong>',
+                replacement: (match, g1) => this.modifyCodeFenceForTag('strong', g1),
             },
             {
                 name: 'strikethrough',
@@ -163,8 +163,20 @@ export default class ExpensiMark {
             },
             {
                 name: 'heading1',
-                regex: /^#(?!#) +(.+)$\s*/gm,
-                replacement: '<h1>$1</h1>',
+                regex: /^#(?!#) +(?! )(.+)$(\s*)/gm,
+                replacement: (__, g1, g2) => {
+                    const index = g1.indexOf('<pre>');
+                    if (index === -1) {
+                        return `<h1>${g1}</h1>`;
+                    }
+                    if (index === 0) {
+                        // heading is empty and keep text #
+                        return `# ${g1}${g2}`;
+                    }
+
+                    // heading is not empty and cut after the tag
+                    return `<h1>${g1.substr(0, index)}</h1>${g1.substr(index)}${g2}`;
+                },
             },
             {
                 name: 'newline',
@@ -349,6 +361,33 @@ export default class ExpensiMark {
         return replacedText;
     }
 
+    modifyCodeFenceForTag(tagName, textToCheck) {
+        const openTag = `<${tagName}>`;
+        const closeTag = `</${tagName}>`;
+
+        const textSplitted = textToCheck.split('<pre>');
+        let joinedText = '';
+        for (let i = 0; i < textSplitted.length; i++) {
+            let text = '';
+            if (Str.contains(textSplitted[i], '</pre>')) {
+                const splitText = textSplitted[i].split('</pre>');
+                if (splitText[1].trim() === '') {
+                    // retain spaces, like 'start<pre>code</pre>   <pre>code2</pre>'
+                    text = `<pre>${splitText[0]}</pre>${splitText[1]}`;
+                } else {
+                    // retain leading and trailing spaces, like 'start<pre>code></pre>    middle    <pre>code2</pre>'
+                    const trimmedText = splitText[1].trim();
+                    text = `<pre>${splitText[0]}</pre>${Str.cutAfter(splitText[1], trimmedText)}${openTag}${trimmedText}${closeTag}${Str.cutBefore(splitText[1], trimmedText)}`;
+                }
+            } else if (textSplitted[i].trim() !== '') {
+                // retain trailing spaces, like 'start     <pre>code</pre>'
+                text = `${openTag}${textSplitted[i].trim()}${closeTag}${Str.cutBefore(textSplitted[i], textSplitted[i].trim())}`;
+            }
+            joinedText += text;
+        }
+        return joinedText;
+    }
+
     /**
      * Checks matched URLs for validity and replace valid links with html elements
      *
```

**Comment by MelvinBot** [comment]
@jliexpensify @roryabraham @aimane-chnaif this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by jliexpensify** [comment]
Not overdue, just waiting on @aimane-chnaif to review some proposals.

**Comment by eh2077** [comment]
Hi @aimane-chnaif , thanks for the feedback.

> if we wanna prevent italic in code block, another complex logic should be added. 


No, we don’t need another logic to fix the italic case. We can reuse function `modifyCodeFenceForTag` for the italic case by 
```typescript
replacement: (match, g1) => this.modifyCodeFenceForTag('em', g1),
```
See below demo video 

https://user-images.githubusercontent.com/117511920/224458795-14ba5889-acdd-4666-a71b-8fa48bd6f378.mov


Hmm, I think the function `modifyCodeFenceForTag` I'm going to add is generic and readable to handle bold and italic cases. We already have several similar functions to manipulate text in `ExpensiMark.js`, like [modifyTextForQuote](https://github.com/Expensify/expensify-common/blob/8451a7d7b9ebe1b7d127cc35793c95b37df2d0e7/lib/ExpensiMark.js#L553). 
So I don't think the solution I proposed is obviously anti-pattern to the existing codebase.

cc @roryabraham

**Comment by situchan** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Various text styles (bold, italic, etc) are applied to code block

### What is the root cause of that problem?
When code block is inside formatting markdown (i.e. # or *), it is treated as normal text and text styles are applied.
Both frontend and backend issues are here.
i.e.
````*Test1 ```code``` Test2*````
In frontend, this is converted to `<strong>Test1 <pre>code</pre> Test2</strong>` (good)
But in backend, it is converted to `<strong>Test1 </strong><pre><strong>code</strong></pre><strong> Test2</strong>` (bad)

````# Test1 ```code``` Test2````
In frontend, this is converted to `<h1>Test1 <pre>code</pre> Test2</h1>` (bad)
But in backend, it is converted to `<h1>Test1 </h1><pre>code</pre> Test2` (good)

### What changes do you think we should make in order to solve the problem?
We can override formatted text style in code block with code block text style.
https://github.com/Expensify/App/blob/a1ca41521c69fa1caf8e0351ef62292238379d63/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js#L31-L47
To achieve this, add custom `pre` styles in customHTMLElementModels of engine provider
```
    pre: defaultHTMLElementModels.div.extend({
        tagName: 'pre',
        mixedUAStyles: {fontStyle: 'normal', fontWeight: 'normal'},
    }),
```

`fontStyle: 'normal', fontWeight: 'normal'` - overrides italic, bold styles

Other markdowns work well but `h1` needs update a bit to prevent applying h1 style to text after code block on the same line:
```diff
            {
                name: 'heading1',
-               regex: /^#(?!#) +(?! )(.+)$\s*/gm,
+               regex: /^# +(?! )((?:(?!<pre>).)+)\s*/gm,
                replacement: '<h1>$1</h1>',
            },
```
Regex explanation:
i.e. ````# Test1 ```code``` Test2````
Currently, h1 style is applied to all including code block and `Test2` but as per the expected behavior, h1 style should be applied to `Test1` only.
So if string includes `<pre>`, stop at here and wrap with h1 tag.
Optional: `(?!#)` is redundant because next character requires space. So this is just for optimization.

As explained in root cause section, bold issue should be fixed in backend as well.

**After fix:**

````# Test1 ```code``` Test2````
In frontend, this is converted to `<h1>Test1 </h1><pre>code</pre> Test2` (to fix)
No change requires in backend

````*Test1 ```code``` Test2*````
In frontend, this is converted to `<strong>Test1 <pre>code</pre> Test2</strong>` (no change)
In backend, keep frontend html, not changing to `<strong>Test1 </strong><pre><strong>code</strong></pre><strong> Test2</strong>` (to fix).
Same applies to italic (`_`, `<em>`)

Demo:

https://user-images.githubusercontent.com/108292595/224970806-69c02823-99fd-49e8-8dbb-ae9ad15aa635.mov

All work well in offline mode. But when online, bold and italic still applied to code block until backend fix.

**Comment by MelvinBot** [comment]
@jliexpensify @roryabraham @aimane-chnaif this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @aimane-chnaif is eligible for the Internal assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
@jliexpensify, @roryabraham, @aimane-chnaif Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by jliexpensify** [comment]
Not overdue - @roryabraham , we need to drop this back to $2000 right? Also, looks like we're awaiting your feedback?

**Comment by roryabraham** [comment]
Sorry for the delay:

> leave code block inside tag as is but force apply text style inside code block in custom render engine provider
i.e. <strong>Test1 <pre>code</pre> Test2</strong>

Yes, I agree this sounds like the better solution.

**Comment by roryabraham** [comment]
@situchan Once you have a draft PR of the front-end changes up, I can work on the back-end changes.

**Comment by situchan** [comment]
@situchan @roryabraham [PR](https://github.com/Expensify/expensify-common/pull/513) in external repo for fixing h1 is ready for review.
Now I will raise draft PR in app repo to fix bold, italic issues as well. For now, I will point commit hash to [bb9e43cb61399c124f286928d92f915c38526dc8](https://github.com/Expensify/expensify-common/pull/513/commits/bb9e43cb61399c124f286928d92f915c38526dc8). This can be updated after above PR is merged.

**Comment by situchan** [comment]
@aimane-chnaif @roryabraham [PR](https://github.com/Expensify/expensify-common/pull/513) in external repo for fixing h1 is ready for review.
Now I will raise draft PR in app repo to fix bold, italic issues as well. For now, I will point commit hash to [bb9e43cb61399c124f286928d92f915c38526dc8](https://github.com/Expensify/expensify-common/pull/513/commits/bb9e43cb61399c124f286928d92f915c38526dc8). This can be updated after above PR is merged.

**Comment by situchan** [comment]
@roryabraham app [PR](https://github.com/Expensify/App/pull/16437) is ready to test backend changes.

**Comment by jliexpensify** [comment]
@roryabraham - can I confirm that we're leaving this job as $4k, as per this [comment](https://github.com/Expensify/App/issues/15193#issuecomment-1476857854)?

**Comment by jliexpensify** [comment]
Invited @mananjadhav (reporting) and @aimane-chnaif . I can't find you in upworks @situchan , can you apply for the job or link your profile?

**Comment by jliexpensify** [comment]
Ugh, sorry haha

**Comment by situchan** [comment]
> I can't find you in upworks @situchan , can you apply for the job or link your profile?

@jliexpensify applied, thank you.

Just FYI: I put this issue on my radar and started research after price bump to 4k on [slack](https://expensify.slack.com/archives/C049HHMV9SM/p1678723264394879) since I love to work with markdown, regex 🙂

**Comment by roryabraham** [comment]
> @roryabraham - can I confirm that we're leaving this job as $4k, as per this https://github.com/Expensify/App/issues/15193#issuecomment-1476857854?

@jliexpensify yes, confirmed

**Comment by roryabraham** [comment]
I'm working on the back-end piece here, and it's caused by [HTMLPurifier](https://github.com/ezyang/htmlpurifier). It seems like we're using mostly the default configs of this library, but I don't know much about it so I need to dig in to see how to adjust this behavior:

```php
$config = HTMLPurifier_Config::createDefault();
// ... some other configs that aren't relevant
$purifier = new HTMLPurifier($config);
$sanitizedComment = $purifier->purify('<strong>Test1 <pre>code</pre> Test2</strong>');
// Result is '<strong>Test1 </strong><pre><strong>code</strong></pre><strong> Test2</strong>'
```

Feel free to chime in if you know more about this or figure it out.

**Comment by roryabraham** [comment]
Ok, so after digging in some more, I see why this is happening. HTMLPurifier helps ensure that HTML is well-formed. If you look into the HTML specifications for the [`<strong>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/strong), you'll see that it lists [phrasing content](https://developer.mozilla.org/en-US/docs/Web/HTML/Content_categories#phrasing_content) as it's permitted content.

Phrasing content includes `<code>`, but not `<pre>`. So `HTMLPurifier` is seeing that we're putting an un-permitted child inside a `<strong>` tag, and "fixing" that for us.

**Comment by roryabraham** [comment]
Not on HOLD, but we're discussing how best to proceed [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1680027731102599)

**Comment by roryabraham** [comment]
We decided to go a different direction with the expected behavior here, because we realized that the expected behavior here was promoting use of invalid HTML, which we don't want.

So this case will stay the same as the accepted proposal:

`# Test1 ```code``` Test2` => `<h1>Test1 </h1><pre>code</pre> Test2`

But this case will change:

`*Test1 ```code``` Test2*` => `*Test1 <pre>code</pre> Test2*`

Said another way, we won't allow the user to wrap block elements in inline elements, because that doesn't really make sense.

**Comment by eh2077** [comment]
Hi @roryabraham, will you consider [my proposal](https://github.com/Expensify/App/issues/15193#issuecomment-1451252235) to support this case `*Test1 ```code``` Test2*`?

It rewrites
```
*Test1 ```code``` Test2*
```
to
````
*Test1* ```
code
```
 *Test2*
````
to avoid strange use of html.

See demo video https://github.com/Expensify/App/issues/15193#issuecomment-1453198340
<img width="577" alt="image" src="https://user-images.githubusercontent.com/117511920/229190399-0bdaa760-32bf-4bf6-b548-53aa3e2ac1f3.png">

**Comment by MelvinBot** [comment]
@jliexpensify, @roryabraham, @aimane-chnaif, @situchan Huh... This is 4 days overdue. Who can take care of this?

**Comment by MelvinBot** [comment]
@jliexpensify, @roryabraham, @aimane-chnaif, @situchan Eep! 4 days overdue now. Issues have feelings too...

**Comment by jliexpensify** [comment]
Not overdue, but @roryabraham - looks like you have a [comment ](https://github.com/Expensify/App/issues/15193#issuecomment-1492350307)to review?

**Comment by roryabraham** [comment]
@eh2077 sorry, at this point we have already hired someone else for the job and they've had their PR merged. We've also changed our proposal guidelines and expect the [proposal template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md) to be used for every proposal going forward. Thanks!

https://github.com/Expensify/App/pull/16437 has been merged and was deployed to staging ~45 minutes ago.

**Comment by eh2077** [comment]
@roryabraham Got it and thanks for the reply!

**Comment by jliexpensify** [comment]
Just making a note for payment for the next B0 assignee - once deployed to prod - as I am OOO starting from tomorrow:

- $4000 to @situchan 
- $4000 to @aimane-chnaif 
- $250 to @gadhiyamanan

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by jliexpensify** [comment]
Hi @maddylewis - I'm OOO so assigning to you for the payment + list ([here's the amounts](https://github.com/Expensify/App/issues/15193#issuecomment-1504485703))

Upworks job is here: https://www.upwork.com/jobs/~014c5422ceeae647ae

**Comment by maddylewis** [comment]
paid!

**Comment by maddylewis** [comment]
well, dammit.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.99-6 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/16437


If no regressions arise, payment will be issued on 2023-04-20. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@aimane-chnaif / @roryabraham] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@aimane-chnaif / @roryabraham] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@aimane-chnaif / @roryabraham] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@maddylewis] Determine if we should create a regression test for this bug.
- [ ] [@aimane-chnaif] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@maddylewis] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by gadhiyamanan** [comment]
@maddylewis  please close the upwork contract
 Thanks!

**Comment by maddylewis** [comment]
okay, so it sounds like @aimane-chnaif and @situchan are both eligible for the 50% merge bonus. is that correct? 

would the bonus be 50% of the final payment or the original payment (ie, $500 or $2000)?

**Comment by maddylewis** [comment]
offers sent

**Comment by gadhiyamanan** [comment]
@maddylewis just a reminder to close the [Upwork contract](https://www.upwork.com/nx/wm/workroom/33129236/overview)

**Comment by maddylewis** [comment]
payments sent - i believe we are all set here.
